core.py

```diff
@@ -620,5 +622,47 @@ def loop(env, minds, vision=Vision(), n_episodes=1, max_steps=-1, policy='determ
     callbacks_list.on_loop_end(False)
     return history.history
 
+
+class Worker(object):
+    """Loop worker."""
+
+    def __init__(self, env_factory, mind_factory, loop_kwargs):
+        self.env_factory = env_factory
+        self.mind_factory = mind_factory
+        self.loop_kwargs = loop_kwargs
+
+        self.env = None
+
+    def __call__(self, job):
+        if self.env is None:
+            self.env = self.env_factory()
+
+        mind = self.mind_factory(job)
+        return loop(self.env, mind, **self.loop_kwargs)
```

Comment:
w jaki sposób jest zapewnione że różne procesy nie będą pisały jednocześnie do jednego pliku?

Reply:
Gdzie tu jest pisanie do jednego pliku?
Jeśli chodzi Ci o callbacki, to te muszą same zadbać, żeby być thread-safe. Więc w kontekście `StoreTrajectories2HDF` trzeba będzie przygotować wersję thread-safe jeśli będziemy z tego korzystać w AlphaZero.

Reply:
Masz rację, tu nie ma pisania do pliku, ale nadal widzę problem. Współdzielenie tych samych callbacków w obrębie różnych procesów (czy wątków?) jest nietypowe. 

W sensie, że normalnie byśmy uruchamiali kilka pętli i za każdym razem:

* inicjalizujemy stan początkowy przed rozpoczęciem pętli
* zbieramy dane podczas kolejnych kroków
* zapisujemy stan po zakończeniu pętli

Tutaj mamy współbieżne działanie więc:

* w losowej kolejności inicjalizujemy stan początkowy
* w losowej kolejności zbieramy dane podczas kolejnych ruchów - nie możemy określić jakie dane przyszły w poprzednim kroku, bo nie wiemy z której pętli przyszły dane + problem wyścigów przy dostępach do pól klasy
* w losowej kolejności zapisujemy stan po zakończeniu pętli

Do tego dochodzi to że takie zapewnianie thread-safe trzeba zrobić w **każdym** callbacku używanym przy hrl.pool. Może warto by stworzyć jakąś klasę SharedCallback, która by dodała thread-safe do zwykłego Callbacka?

Reply:
Chyba trochę nie rozumiesz co tu się dzieje :D Każdy worker jest oddzielnym PROCESEM, nie wątkiem. Procesy jak wiemy są odizolowane i nie mogą ze sobą dzielić tak po prostu pamięci, obiektów itp. Każdy worker (najprawdopodobniej) dostaje kopie runtime-u, dzięki temu wszystkie moduły itp. są dostępne cały czas. Ale wszystkie obiekty są kopiami w każdym procesie, zatem jak masz czterech worker-ów, to każdy działa na swojej kopii callbacka i Mind-a ;)